src/Features/CSharp/Portable/UsePatternMatching/CSharpAsAndNullCheckDiagnosticAnalyzer.cs

```diff
@@ -175,225 +138,134 @@ private void SyntaxNodeAction(SyntaxNodeAnalysisContext syntaxContext)
+                        if (!TryFindVariableDeclarator(semanticModel, identifier, out localSymbol, out declarator))
+                        {
+                            break;
+                        }
+
+                        var initializerValue = declarator.Initializer?.Value;
+                        if (!initializerValue.IsKind(SyntaxKind.AsExpression, out asExpression))
+                        {
+                            break;
+                        }
 
-                    variableDeclarator = declarator;
-                    asExpression = initializerValue;
-                    return true;
-                }
+                        return true;
+                    }
 
                 case SyntaxKind.SimpleAssignmentExpression:
-                {
-                    // T x;
-                    // if ((x = e as T) != null) F(x);
-                    var assignment = (AssignmentExpressionSyntax)operand;
-                    if (!assignment.Right.IsKind(SyntaxKind.AsExpression) ||
-                        !assignment.Left.IsKind(SyntaxKind.IdentifierName))
                     {
-                        break;
-                    }
+                        // T x;
+                        // if ((x = e as T) != null) F(x);
+                        var assignment = (AssignmentExpressionSyntax)operand;
+                        if (!assignment.Right.IsKind(SyntaxKind.AsExpression, out asExpression) ||
+                            !assignment.Left.IsKind(SyntaxKind.IdentifierName, out IdentifierNameSyntax identifier))
+                        {
+                            break;
+                        }
+
+                        if (!TryFindVariableDeclarator(semanticModel, identifier, out localSymbol, out declarator))
+                        {
+                            break;
+                        }
 
-                    var identifier = (IdentifierNameSyntax)assignment.Left;
-                    var declarator = TryFindVariableDeclarator(identifier, targetStatement, parentBlock);
-                    if (declarator == null)
-                    {
-                        break;
+                        return true;
                     }
-
-                    variableDeclarator = declarator;
-                    asExpression = assignment.Right;
-                    return true;
-                }
             }
 
-            variableDeclarator = null;
+            declarator = null;
             asExpression = null;
+            localSymbol = null;
             return false;
         }
 
-        private static VariableDeclaratorSyntax TryFindVariableDeclarator(
-            IdentifierNameSyntax identifier, StatementSyntax targetStatement, BlockSyntax parentBlock)
+        private static bool TryFindVariableDeclarator(
+            SemanticModel semanticModel,
+            IdentifierNameSyntax identifier,
+            out ILocalSymbol localSymbol,
+            out VariableDeclaratorSyntax declarator)
```

Comment:
>VariableDeclaratorSyntax [](start = 16, length = 24)

`VariableDeclaratorSyntax` can appear in multiple contexts, not just local declaration statements. For instance: `for` statements, `using`, `lock`, and possibly others.
I didn't see where we guard against that.

Reply:
In `CanSafelyConvertToPatternMatching` we skip over that to get to the parent.